solutions/silver/usaco-645.mdx

```diff
@@ -2,159 +2,180 @@
 using namespace std;
 
-const int MX = 5e4;
-int n, x[MX], y[MX], ar[MX];
+using ll = long long;
 
-bool cmp(int a, int b) {
-	if (x[a] == x[b]) return y[a] < y[b];
-	return x[a] < x[b];
-}
+int N;
+ll ans = 0;
+vector<pair<int, int>> cows;
 
-bool cmp1(int a, int b) {
-	if (y[a] == y[b]) return x[a] < x[b];
-	return y[a] < y[b];
+void search() {
+	sort(cows.begin(), cows.end());
+
+	vector<pair<int, int>> pref(N), suf(N);
+
+	auto upd = [](pair<int, int> a, int b) -> pair<int, int> {
+		return {min(a.first, b), max(a.second, b)};
+	};
+
+	// compute prefix/suffix minimums and maximums
+	pref[0] = {cows[0].second, cows[0].second};
+	for (int i = 1; i < N; ++i) { pref[i] = upd(pref[i - 1], cows[i].second); }
+
+	suf[N - 1] = {cows[N - 1].second, cows[N - 1].second};
+	for (int i = N - 2; i >= 0; i--) {
+		suf[i] = upd(suf[i + 1], cows[i].second);
+	}
+
+	// initial area with one enclosure
+	ll area = (ll)(cows.back().first - cows.front().first) *
+	          (pref.back().second - pref.back().first);
+
+	ll best = LLONG_MAX;
+	for (int i = 0; i < N - 1; ++i) {
+		// if it's possible to split
+		if (cows[i].first != cows[i + 1].first) {
+			ll first_rect = (ll)(cows[i].first - cows[0].first) *
+			                (pref[i].second - pref[i].first);
+			ll second_rect = (ll)(cows.back().first - cows[i + 1].first) *
+			                 (suf[i + 1].second - suf[i + 1].first);
+
+			best = min(best, first_rect + second_rect);
+		}
+	}
+
+	ans = max(ans, area - best);
 }
 
 int main() {
 	freopen("split.in", "r", stdin);
 	freopen("split.out", "w", stdout);
-	cin >> n;
-	multiset<int> allX, allY, leftX, leftY, rightX, rightY;
-	for (int i = 0; i < n; i++) {
-		ar[i] = i;
-		cin >> x[i] >> y[i];
-		allX.insert(x[i]);
-		allY.insert(y[i]);
-	}
-	sort(ar, ar + n, cmp);
-	int x1 = *allX.rbegin(), x2 = *allX.begin(), y1 = *allY.rbegin(),
-	    y2 = *allY.begin();
-	long long area = (x1 - x2) * 1LL * (y1 - y2);  // original area
-	long long sol = 0;                             // maximum difference
-
-	for (int k = 0; k <= 1; k++) {
-		rightX = allX;
-		rightY = allY;
-		leftX.clear();
-		leftY.clear();
-		for (int i = 0; i < n; i++) {
-			int j = ar[i];
-			long long areaL = 0, areaR = 0;
-
-			leftX.insert(x[j]);
-			leftY.insert(y[j]);
-			rightX.erase(rightX.find(x[j]));
-			rightY.erase(rightY.find(y[j]));
-
-			x1 = *leftX.begin();
-			y1 = *leftY.begin();
-			x2 = *leftX.rbegin();
-			y2 = *leftY.rbegin();
-			areaL = (x2 - x1) * 1LL * (y2 - y1);
-			if (!rightX.empty()) {
-				x1 = *rightX.begin();
-				y1 = *rightY.begin();
-				x2 = *rightX.rbegin();
-				y2 = *rightY.rbegin();
-				areaR = (x2 - x1) * 1LL * (y2 - y1);
-			}
-			sol = max(sol, area - (areaR + areaL));
-		}
-		sort(ar, ar + n, cmp1);
-	}
-	cout << sol << '\n';
+
+	cin >> N;
+	cows.resize(N);
+
+	for (pair<int, int> &cow : cows) cin >> cow.first >> cow.second;
+
+	// try all splits on x axis
+	search();
+
+	for (pair<int, int> &cow : cows) swap(cow.first, cow.second);
```

Comment:
braces

Reply:
not fixed?

Reply:
im so sorry i thought i did it :<<
thanks